File: MovieTicketBooking/Providers/InMemorySeatLockProvider.cs
using MovieTicketBooking.Model;
using MovieTicketBooking.Services;

namespace MovieTicketBooking.Providers
{
    public class InMemorySeatLockProvider : ISeatLockProvider
    {
        Dictionary<Show, Dictionary<Seat, SeatLock>> _locks = new();
        int _lockTimeout;
        public InMemorySeatLockProvider(int lockTimeout)
        {
            _lockTimeout = lockTimeout;
        }
        public List<Seat> GetLockedSeats(Show show)
        {
            List<Seat> seats = new();
            if(!_locks.ContainsKey(show))
            {
                return seats;
            }
            var allSeats = _locks.GetValueOrDefault(show).Keys;
            foreach(var seat in allSeats)
            {
                if(isSeatLocked(show, seat))
                {
                    seats.Add(seat);
                }
            }
            return seats;
        }

        public void LockSeats(Show show, List<Seat> seats, string user)
        {
            foreach (Seat seat in seats)
            {
                if (isSeatLocked(show, seat))
                {
                    throw new Exception("seat is locked already");
                }
            }
            foreach (Seat seat in seats)
            {
                lockSeat(user, seat, show, _lockTimeout);
            }
        }

        private void lockSeat(string user, Seat seat, Show show, int timeoutInSeconds)
        {
            SeatLock seatLock = new SeatLock(user, seat, show, timeoutInSeconds, DateTime.Now);
            if (!_locks.ContainsKey(show))
            {
                _locks.Add(show, new Dictionary<Seat, SeatLock>());
            }
            _locks.GetValueOrDefault(show).Add(seat, seatLock);
        }
        public void UnlockSeat(Show show, List<Seat> seats, string user)
        {
            foreach (Seat seat in seats)
            {
                if(ValidateLock(show, seat, user))
                {
                    unlockSeat(show, seat);
                }
            }
        }

        public bool ValidateLock(Show show, Seat seat, string user)
        {
            return isSeatLocked(show, seat) &&
            _locks.GetValueOrDefault(show).GetValueOrDefault(seat).LockedBy.Equals(user);
        }

        private void unlockSeat(Show show, Seat seat)
        {
            if (!_locks.ContainsKey(show))
            {
                return;
            }
            _locks.GetValueOrDefault(show).Remove(seat);
        }

        private bool isSeatLocked(Show show, Seat seat)
        {
            return _locks.ContainsKey(show) &&
            _locks.GetValueOrDefault(show).ContainsKey(seat) &&
            !_locks.GetValueOrDefault(show).GetValueOrDefault(seat).isLockExpired();
        }
    }
}

File: MovieTicketBooking/Movie Booking Java/src/main/java/org/example/providers/InMemorySeatLockProvider.java
package org.example.providers;

import com.google.common.collect.ImmutableList;
import lombok.NonNull;
import org.example.exceptions.SeatNotAvailableException;
import org.example.model.Seat;
import org.example.model.SeatLock;
import org.example.model.Show;

import java.time.LocalDateTime;
import java.util.*;

public class InMemorySeatLockProvider implements SeatLockProvider {

    private final Map<Show, Map<Seat, SeatLock>> seatLocks;
    private final Integer lockTimeOut;

    public InMemorySeatLockProvider(Integer lockTimeOut) {
        seatLocks = new HashMap<>();
        this.lockTimeOut = lockTimeOut;
    }

    @Override
    public void lockSeat(final @NonNull List<Seat> seats, final @NonNull Show show, final @NonNull String userId) {

        for (Seat seat : seats) {
            if (isSeatLocked(show, seat)) {
                throw new SeatNotAvailableException("seat not available at this moment!");
            }
        }

        for (Seat seat : seats) {
            lockSeat(seat, show, userId);
        }

    }

    @Override
    public void unlockSeat(List<Seat> seats, Show show, String userId) {

        for (Seat seat : seats) {
            if (validateSeatLock(seat, show, userId)) {
                removeSeat(seat, show, userId);
            }
        }
    }

    @Override
    public boolean validateSeatLock(Seat seat, Show show, String userId) {
        return isSeatLocked(show, seat) && this.seatLocks.get(show).get(seat).getLockedBy().equals(userId);
    }

    @Override
    public List<Seat> getLockedSeats(Show show) {

        if (!seatLocks.containsKey(show)) {
            return ImmutableList.of();
        }

        final List<Seat> lockedSeats = new ArrayList<>();

        for (Map.Entry<Seat, SeatLock> seatLockEntry : seatLocks.get(show).entrySet()) {

            Seat seat = seatLockEntry.getKey();

            if (isSeatLocked(show, seat)) {
                lockedSeats.add(seatLockEntry.getKey());
            }
        }

        return lockedSeats;
    }

    private void lockSeat(final Seat seat, final Show show, final String userId) {


        if (!this.seatLocks.containsKey(show)) {
            this.seatLocks.put(show, new HashMap<>());
        }
        SeatLock seatLock = new SeatLock(seat, show, LocalDateTime.now(), lockTimeOut, userId);
        this.seatLocks.get(show).put(seat, seatLock);
    }

    private void removeSeat(final Seat seat, final Show show, final String userId) {

        if (!seatLocks.containsKey(show)) {
            return;
        }
        this.seatLocks.get(show).put(seat, null);
    }

    private boolean isSeatLocked(final Show show, final Seat seat) {

        if (seatLocks.containsKey(show)) {

            Map<Seat, SeatLock> seatLockMap = seatLocks.get(show);
            if (Objects.nonNull(seatLockMap.get(seat))) {
                return !seatLockMap.get(seat).isLockExpired();
            }
        }

        return false;
    }

}
